File: leetcode/2 add two numbers/Main.java
import java.util.*;
import java.io.*;
import java.math.*;
import java.util.regex.*;
import static java.lang.System.*;
public class Main{

    static long start, stop, mod = (long) 1e9 + 7L;
    static void debug(Object... a){
        err.println(Arrays.deepToString(a));
    }
    static void starttime(){start = currentTimeMillis();}
    static void stoptime(){stop = currentTimeMillis();}
    static void gettime(){err.println(((double)(stop - start) / 1000.0) + " seconds");}

    static class FastReader{
        StringTokenizer st;
        BufferedReader br;
        FastReader(){
            br = new BufferedReader(new InputStreamReader(in));
        }
        String n(){
            while(st == null || !st.hasMoreElements()){
                try{
                    st = new StringTokenizer(br.readLine());
                }
                catch(Exception e){
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int ni(){
            return Integer.parseInt(n());
        }
        float nf(){
            return Float.parseFloat(n());
        }

        double nd(){
            return Double.parseDouble(n());
        }
        long nl(){
            return Long.parseLong(n());
        }
        String ns(){
            String k="";
            try{
                k = br.readLine();
            }
            catch(Exception e){

                e.printStackTrace();
            }
            return k;

        }
    }
    static class pair<First, Second>{
		First first;
		Second second;
		pair(First first,Second second){
			this.first = first;
			this.second = second;
		}
		public First getFirst(){
			return this.first;
		}
		public Second getSecond(){
			return this.second;
		}
		@Override
		public String toString(){
			return (first + " " + second);
		}
		@Override
		public boolean equals(Object o){
			if(this == o)
				return true;
			if(o == null || getClass() != o.getClass())
				return false;
			pair<?,?> pair = (pair<?, ?>) o;
			if(!first.equals(pair.first)) return false;
			return second.equals(pair.second);
		}
		@Override
		public int hashCode(){
			return 31 * first.hashCode() + second.hashCode();
		}
	}
    
    public static void main(String args[]) throws IOException{
        FastReader in = new FastReader();
        	starttime();
            solve(in);
            stoptime();
            //gettime();
    }
    static class Node{
        int data;
        Node next;
        Node temp;
        int length;
        void add(int data){
            if(temp == null){
                temp = this;
                this.data = data;
            }
            else{
                temp.next = new Node();
                temp.next.data = data;
                temp = temp.next;
            }
            length++;
        }
        void display(){
            Node temp = this;
            while(temp != null){
                out.print(temp.data + " ");
                temp = temp.next;
            }
            out.println();
        }    
    }
    static Node addTwoNumbers(Node first, Node second){
        Node temp = new Node();
        int sum = 0;
        int carry = 0;
        while(first != null || second != null){
            if(first != null){
                carry += first.data;
                first = first.next;
            }
            if(second != null){
                carry += second.data;
                second = second.next;
            }
            temp.add(carry % 10);
            carry /= 10;
            if(first == null && second == null) break;
        }
        if(carry > 0) temp.add(carry);
        return temp;
    }
    static void solve(FastReader in){
        Node first = new Node();
        first.add(9);
        first.add(9);
        first.display();
        Node second = new Node();
        second.add(9);
        second.display();
        addTwoNumbers(first, second).display();
    }
}

File: leetcode/2 add two numbers/Main.py
class List:
    class Node:
        def __init__(self, previous, data, next):
            self.data = data
            self.previous = previous
            self.next = next
        def __str__(self):
            return "[" + str(self.data) + "]"

    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def getHead(self):
        return self.head
    
    def getTail(self):
        return self.tail

    def addLast(self, data)->None:
        last = self.tail
        newnode = self.Node(last, data, None)
        self.tail = newnode
        if last is None:
            self.head = newnode
        else:
            last.next = newnode
        self.size += 1

    def addFirst(self, data)-> None:
        first = self.head
        newnode = self.Node(None, data, first)
        self.head = newnode
        if first is None:
            self.tail = newnode
        else:
            first.previous = newnode
        self.size += 1

    def __str__(self):
        temp = self.head
        answer = "["
        while temp != None:
            answer += str(temp.data)
            if temp.next is not None:
                answer += ", "
            temp = temp.next
        return answer + "]"
    def __len__(self):
        return self.size
        

def addToNumbers(first:List, second:List)-> List:
    carry = 0
    temp = List()
    first = first.getTail()
    second = second.getTail()
    while first is not None or second is not None:
        if first is not None:
            carry += first.data
            first = first.previous
        if second is not None:
            carry += second.data
            second = second.previous
        temp.addFirst(carry % 10)
        carry //= 10
    if carry > 0:
        temp.addFirst(carry)
    return str(temp)
        

if __name__ == '__main__':
    first = List()
    first.addFirst(1)
    first.addFirst(1)
    first.addFirst(1)

    second = List()
    second.addFirst(2)
    second.addFirst(3)

    print(addToNumbers(first, second))
